plutus-tx/src/PlutusTx/Builtins/Class.hs

```diff
@@ -149,8 +151,67 @@ type instance BuiltinRep (a,b) = BuiltinPair (BuiltinRep a) (BuiltinRep b)
 instance (FromBuiltin a, FromBuiltin b) => FromBuiltin (a,b) where
     {-# INLINABLE fromBuiltin #-}
     fromBuiltin p = (fromBuiltin $ fst p, fromBuiltin $ snd p)
+instance ToBuiltin (BuiltinData, BuiltinData) where
+    {-# INLINABLE toBuiltin #-}
+    toBuiltin (d1, d2) = mkPairData d1 d2
 
 type instance BuiltinRep [a] = BuiltinList (BuiltinRep a)
 instance FromBuiltin a => FromBuiltin [a] where
     {-# INLINABLE fromBuiltin #-}
-    fromBuiltin l = ifThenElse (null l) [] (fromBuiltin (head l):fromBuiltin (tail l))
+    fromBuiltin = builtinListToList
+
+{- Note [Specializing recursive FromBuiltin instances]
+Recursive instances of `FromBuiltin`, such as the one for lists, are hard for the plugin to deal with.
+They have a superclass constraint which contains functions in terms of an *unknown* type family application
+(i.e. we have `fromBuiltin :: BuiltinRep a -> a`, where `BuiltinRep a` is unknown).
+The plugin requires all type family applications to reduce away, so this is a no go.
```

Comment:
Would it make things easier not to use a type family and use a multi-parameter type class with a functional dependency instead?

Reply:
Hmmm, maybe that would work indeed. Let me try that, it might be less fragile.

Reply:
That seems to require `UndecidableInstances`: GHC is unhappy with things like this:
```
instance (FromBuiltin arep a, FromBuiltin brep b) => FromBuiltin (BuiltinPair arep brep) (a,b) where
```
It doesn't believe that `BuiltinPair arep brep` determines `(a,b)`. I haven't used functional dependencies much, I don't know if there's a trick here.

Reply:
Also it was nice that I at least got the generic instances for `BuiltinRep [a]` etc, whereas this way I have to write out the rep tyep for the specialized instances, but that's okay.

Reply:
But it does seem to work. So I guess the question is: which is worse, `UndecidableInstances` or this fragile specialization crap? I think the specialization stuff is worse, it relies on far more details about how GHC generates code. So I think the fundeps approach is a winner.

Reply:
> `UndecidableInstances`

Just enable it whenever GHC asks for it and [don't worry about it](https://www.reddit.com/r/haskell/comments/5zjwym/when_is_undecidableinstances_okay_to_use/deynhlx). It's a perfectly benign extension despite it sounding alarming. If GHC was unable to satisfy the functional dependency, you'd get a different error that you couldn't dispose that easily (you could "solve" it by having `instance C a => C a`, but GHC wouldn't tell you about that hack).